Design for OpenID Connect (#4775)
nginxinc/nginx-gateway-fabric: docs/proposals/oidc.md

	//
	// +optional
	// +kubebuilder:validation:Pattern=`^(?:http?:\/\/)?[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?(?:\.[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?)*(?::\d{1,5})?$`
	FrontChannelLogoutURI *string `json:"frontChannelLogoutURI,omitempty"`

	// TokenHint adds the id_token_hint argument to the Provider's Logout Endpoint.
	// Some OpenID Providers require this.
	// Directive: https://nginx.org/en/docs/http/ngx_http_oidc_module.html#logout_token_hint
	// NGINX Default: false
	//
	// +optional
	TokenHint *bool `json:"tokenHint,omitempty"`
}
```

For simplicity, only one OIDC provider can be configured at this time. To set up authentication with an OpenID Provider, you must specify the provider name, issuer URL, client ID, and client secret. The AuthenticationFilter must be attached to a route that uses a TLS listener in terminate mode, with the appropriate certificates provided as a listener secret. This is required because the `redirect_uri` must use HTTPS. When the OpenID Provider redirects users back to NGINX after authentication, the request is made over HTTPS, requiring NGINX to have TLS configured.

> TLS is required for secure communication between the data plane and the OpenID Provider. To verify TLS connections, specify a CA bundle with the appropriate CN/SAN using the `caCertificateRefs` field of the AuthenticationFilter CRD; if omitted, the system CA will be used by default. Additionally, the server certificate and key for authenticating to the OpenID Provider must be configured via the `certificateRef` field on the Gateway's TLS listener, which should be attached to the Route containing the OIDC filter.

Reviewer: ```suggestion
TLS is required for secure communication between the data plane and the OpenID Provider. To verify TLS connections, specify a CA bundle with the appropriate CN/SAN using the `caCertificateRefs` field of the AuthenticationFilter CRD; if omitted, the system CA will be used by default. Additionally, the AuthenticationFilter must be attached to a route using an HTTPS listener so that the redirect_uri callback from the IdP can be served over HTTPS. The Gateway listener's certificateRef provides the TLS certificate for this incoming connection.
```

An authenticationFilter with complete OIDC configuration would look like:

```yaml
apiVersion: gateway.nginx.org/v1alpha1
kind: AuthenticationFilter
metadata:
  name: keycloak-oidc-auth
  namespace: default
spec:
  type: OIDC
  oidc:
    issuer: "https://keycloak.example.com/realms/my-realm"
    clientID: nginx-gateway
    clientSecret:
      name: oidc-client-secret
      namespace: default

    caCertificateRefs:

>       name: oidc-ca-cert

Reviewer: ```suggestion
    caCertificateRefs:
      - name: oidc-ca-cert
```

    session:
      cookieName: "MY_OIDC_SESSION"
      timeout: "4h"

    logout:
      uri: "/logout"
      postLogoutURI: "/logged-out"
      frontChannelLogoutURI: "/frontchannel-logout"
      tokenHint: true
---
# clientSecret references a Kubernetes Secret containing the OIDC client secret
# The OpenID Provider uses this to verify that token requests come from a legitimate client
apiVersion: v1
kind: Secret
metadata:
  name: oidc-client-secret
  namespace: default
type: Opaque
stringData:
  client-secret: "secret-value"
---
# CA bundle used to verify the IdP's certificate when NGINX connects to it.
apiVersion: v1
kind: Secret
metadata:
  name: oidc-ca-cert
  namespace: default
type: Opaque
stringData:
  ca.crt: |
    -----BEGIN CERTIFICATE-----
    MIIDXTCCAkWgAwIBAgIJAJC1HiIAZAiUMA0Gcg...
    -----END CERTIFICATE-----
---
apiVersion: v1
kind: Secret
metadata:
  name: oidc-crl
  namespace: default
type: Opaque
stringData:
  ca.crl: |
    -----BEGIN X509 CRL-----
    MIIBYDCBygIBATANBgkqhkiG9w0BAQsFADA...
    -----END X509 CRL-----
```

Generating the appropriate configuration requires a TLS listener configured with a certificate and key for verification.

> An example would be a Gateway that defines an HTTPS listener and references a Secret for authenticating to the OpenID provider.

Reviewer: Actually, we could just remove these comments altogether, as it's already called out in the Route comment below

- The client exchanges the code for an ID Token and Access Token via a back-channel request.
- The Authorization Server may authenticate the client (using a Client Secret) before issuing tokens.

The following diagram illustrates the Authorization Code Flow:

```mermaid
sequenceDiagram
    participant User as End-User
    participant Keycloak as Keycloak (Client)
    participant NGF as NGF (Authorization Server)

    Note over Keycloak,NGF: Authorization Code Flow

    Keycloak->>Keycloak: 1. Prepare Authentication Request<br/>(client_id, redirect_uri, scope, state)

    Keycloak->>NGF: 2. Send Authentication Request

    NGF->>User: 3. Present Login Page
    User->>NGF: Submit Credentials
    NGF->>NGF: Validate Credentials

    NGF->>User: 4. Present Consent Screen
    User->>NGF: Grant Authorization

    NGF->>Keycloak: 5. Redirect with Authorization Code<br/>(code, state)

    Keycloak->>NGF: 6. Token Request<br/>(code, client_id, client_secret, redirect_uri)

    NGF->>Keycloak: 7. Token Response<br/>(id_token, access_token, refresh_token)

    Keycloak->>Keycloak: 8. Validate ID Token &<br/>Retrieve Subject Identifier

> 

Reviewer: The flow in this chart is the wrong way around - NGINX prepares and sends the auth request to Keycloak, the user submits the credentials to Keycloak, etc

	// .well-known/openid-configuration endpoint.
	// Directive: https://nginx.org/en/docs/http/ngx_http_oidc_module.html#issuer
	// Examples:
	//   - Keycloak: "https://keycloak.example.com/realms/my-realm"
	//   - Okta: "https://dev-123456.okta.com/oauth2/default"
	//   - Auth0: "https://my-tenant.auth0.com/"
	//
	// +kubebuilder:validation:MinLength=1
	// +kubebuilder:validation:Pattern=`^https://`
	// +kubebuilder:validation:Required
	Issuer string `json:"issuer"`

	// ClientID is the client identifier registered with the OpenID Provider.
	// Directive: https://nginx.org/en/docs/http/ngx_http_oidc_module.html#client_id
	//
	// +kubebuilder:validation:MinLength=1
	// +kubebuilder:validation:Required
	ClientID string `json:"clientID"`

	// The Kubernetes secret which contains the OIDC client secret to be used in the
	// [Authentication Request](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest).
	// Directive: https://nginx.org/en/docs/http/ngx_http_oidc_module.html#client_secret
	//
	// This is an Opaque secret. The client secret should be stored in the key
	// "client-secret".
	// +kubebuilder:validation:Required

> 	ClientSecret v1.SecretObjectReference `json:"clientSecret"`

Reviewer 2: If we aren't currently allowing cross-namespace Secret references, then this should be the LocalObjectReference. Same for all other Secrets that are referenced in here.

	// TrustedCertificateRef references a secret containing trusted CA certificates
	// in PEM format used to verify the certificates of the OpenID Provider endpoints.
	// If not specified, the system CA bundle is used.
	//
	// Directive: https://nginx.org/en/docs/http/ngx_http_oidc_module.html#ssl_trusted_certificate
	// NGINX Default: system CA bundle
	//
	// +optional
	TrustedCertificateRef *v1.SecretObjectReference `json:"trustedCertificateRef,omitempty"`

	// CertificateRevocationList references a key in a Secret containing a certificate
	// revocation list (CRL) in PEM format. This is used to verify that certificates
	// presented by the OpenID Provider endpoints have not been revoked.
	//
	// +optional

> 	CertificateRevocationList *v1.SecretObjectReference `json:"certificateRevocationList,omitempty"`

Reviewer 2: We should define a field name that the crl file is stored in in this Secret, like `oidc.crl` or something along those lines. We also may want it to be a custom Secret type, like we do with JWT and Basic auth.

This probably warrants its own section in this doc to discuss this as well.

Reviewer 2: Or if there's already a k8s standard for CRL Secrets, use that instead. Not sure if there is.

Reviewer 2: Are they CA certificates in the CRL? Or client certificates?

Reviewer 2: Stapling would be defined elsewhere though, right? `ssl_crl` is in the `oidc_provider`, stapling is not. So it doesn't seem like we even need to consider that?

My previous comment was mainly to figure out what the best name is for the Secret field for the CRL contents.

Reviewer 2: Spoke offline, ca.crl is good.

	// When enabled, UserInfo claims are available via $oidc_claim_<name> variables.
    // NGINX Directive: https://nginx.org/en/docs/http/ngx_http_oidc_module.html#userinfo
	// NGINX Default: false
	//
	// +optional
	UserInfo *bool `json:"userInfo,omitempty"`

	// PKCE enables Proof Key for Code Exchange (PKCE) for the authentication flow.
    // If nil, NGINX automatically enables PKCE when the OpenID Provider requires it.
	// Directive: https://nginx.org/en/docs/http/ngx_http_oidc_module.html#pkce
	//
	// +optional
	PKCE *bool `json:"pkce,omitempty"`

	// ExtraAuthArgs sets additional query arguments for the authentication request URL.
	// Arguments are appended with "&". For example: "prompt=consent&audience=api".
	// Directive: https://nginx.org/en/docs/http/ngx_http_oidc_module.html#extra_auth_args
	//
	// +optional
	ExtraAuthArgs map[string]string `json:"extraAuthArgs,omitempty"`

	// OIDCSession configures session management for OIDC authentication.
	//
	// +optional
	OIDCSession *OIDCSessionConfig `json:"oidcSession,omitempty"`

	// OIDCLogout defines the logout behavior for OIDC authentication.
	//
	// +optional

> 	OIDCLogout *OIDCLogoutConfig `json:"oidcLogout,omitempty"`

Reviewer 2: I wouldn't put OIDC in the names of these fields, it's redundant.

```suggestion
	// Session configures session management for OIDC authentication.
	//
	// +optional
	Session *OIDCSessionConfig `json:"session,omitempty"`

	// Logout defines the logout behavior for OIDC authentication.
	//
	// +optional
	Logout *OIDCLogoutConfig `json:"logout,omitempty"`
```

	// ClaimHeaders configures which OIDC claims to forward as headers to the backend.
	//
	// +optional

> 	ClaimHeaders []ClaimHeader `json:"claimHeaders,omitempty"`

Reviewer 2: This is similar to the JWT case, where a user could technically use a RequestHeaderModifier, right?

        location = /coffee {
            auth_oidc keycloak;
            auth_jwt off token=$oidc_id_token;

            proxy_set_header X-User-ID $oidc_claim_sub;
            proxy_set_header X-User-Email $oidc_claim_email;
            proxy_set_header X-User-Name $oidc_claim_name;
            proxy_set_header X-Username $oidc_claim_preferred_username;

            proxy_set_header X-Postal-Code $claim_address_postal_code;
            proxy_set_header X-User-Roles $claim_realm_access_roles;

            proxy_pass http://default_coffee_80$request_uri;
        }
    }
}
```

### Error Handling

1. The OpenID Provider is down - When the OpenID provider is unreachable during initial authentication, NGINX returns a 302 redirect to the OpenID provider's authorization endpoint. The connection failure occurs client-side when the browser attempts to reach the OpenID provider, resulting in a "connection refused" error. NGINX does not validate OpenID providers availability before issuing the redirect.
2. If the referenced `clientSecret` or `trustedCertificateRef` resources don't exist, the AuthenticationFilter should
   report an error condition and the route should not be programmed.

> 3. If the `client-secret` is incorrect for OpenID provider, the redirect to the provider succeeds but the redirect back to NGINX fails and the error is reflected in logs as `token error: unauthorized_client: Invalid client or Invalid client credentials while sending to client`

Reviewer 2: In addition, if attached to a non-HTTPS listener, it should not be accepted.

	//
	// This is an Opaque secret. The client secret should be stored in the key
	// "client-secret".
	ClientSecret v1.LocalObjectReference `json:"clientSecret"`

	// TrustedCACertificateRefs references a secret containing trusted CA certificates
	// in PEM format used to verify the certificates of the OpenID Provider endpoints.
    // The CA certificates must be stored in a key named `ca.crt`.
	// If not specified, the system CA bundle is used.
	//
	// Directive: https://nginx.org/en/docs/http/ngx_http_oidc_module.html#ssl_trusted_certificate
	// NGINX Default: system CA bundle
	//
	// +optional
    // +kubebuilder:validation:MaxItems=8

> 	TrustedCACertificateRefs []v1.LocalObjectReference `json:"trustedCACertificateRefs,omitempty"`

Reviewer 2: I don't think we need Trusted in the name, per consistency with our other fields.
